tests/unit/s2n_key_material_test.c

```diff
@@ -0,0 +1,209 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_prf.h"
+
+S2N_RESULT helper_generate_test_data(struct s2n_blob *test_data)
+{
+    struct s2n_stuffer test_data_stuffer = { 0 };
+    EXPECT_SUCCESS(s2n_stuffer_init(&test_data_stuffer, test_data));
+    for (int i = 0; i < S2N_MAX_KEY_BLOCK_LEN; i++) {
+        RESULT_GUARD_POSIX(s2n_stuffer_write_uint8(&test_data_stuffer, i));
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT helper_validate_key_material(struct s2n_key_material *key_material, struct s2n_blob *test_data_blob,
+        uint8_t mac_size, uint8_t key_size, uint8_t iv_size)
+{
+    /* confirm that the data is copied to key_material */
+    RESULT_ENSURE_EQ(memcmp(test_data_blob->data, key_material->key_block, test_data_blob->size), 0);
+
+    /* test that its possible to access data from s2n_key_material */
+    uint8_t *test_ptr = test_data_blob->data;
+    RESULT_ENSURE_REF(test_ptr);
+    /* client MAC */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_mac.data, mac_size), 0);
+    test_ptr += mac_size;
+    RESULT_ENSURE_REF(test_ptr);
+    /* server MAC */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_mac.data, mac_size), 0);
+    test_ptr += mac_size;
+    RESULT_ENSURE_REF(test_ptr);
+
+    /* client KEY */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_key.data, key_size), 0);
+    test_ptr += key_size;
+    RESULT_ENSURE_REF(test_ptr);
+    /* server KEY */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_key.data, key_size), 0);
+    test_ptr += key_size;
+    RESULT_ENSURE_REF(test_ptr);
+
+    /* client IV */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->client_iv.data, iv_size), 0);
+    test_ptr += iv_size;
+    RESULT_ENSURE_REF(test_ptr);
+    /* server IV */
+    RESULT_ENSURE_EQ(memcmp(test_ptr, key_material->server_iv.data, iv_size), 0);
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* prepare test data */
+    uint8_t test_data[S2N_MAX_KEY_BLOCK_LEN] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(helper_generate_test_data(&test_data_blob));
+
+    /* fuzz s2n_key_material_init with different mac, key, iv sizes */
+    {
+        for (uint8_t mac_size = 0; mac_size < 16; mac_size++) {
+            for (uint8_t key_size = 0; key_size < 32; key_size++) {
+                for (uint8_t iv_size = 0; iv_size < 32; iv_size++) {
```

Comment:
Rather than testing/fuzzing arbitrary values, would it make sense/be cleaner to just do a single loop over `s2n_all_cipher_suites` and test all of the combinations actually in use?

Reply:
s2n_all_cipher_suites is tested in other part of our code so I didnt want to test that here again. Here I wanted to generally test that s2n_key_material can handle all combinations of values independently. For example `0`